simple_diarizer/Spectral_clustering.py

```diff
@@ -0,0 +1,135 @@
+import numpy as np
+import scipy
+from sklearn.cluster import SpectralClustering
+
+# NME low-level operations
+# These functions are taken from the Kaldi scripts.
+
+# Prepares binarized(0/1) affinity matrix with p_neighbors non-zero elements in each row
+def get_kneighbors_conn(X_dist, p_neighbors):
+    X_dist_out = np.zeros_like(X_dist)
+    for i, line in enumerate(X_dist):
+        sorted_idx = np.argsort(line)
+        sorted_idx = sorted_idx[::-1]
+        indices = sorted_idx[:p_neighbors]
+        X_dist_out[indices, i] = 1
+    return X_dist_out
+
+
+# Thresolds affinity matrix to leave p maximum non-zero elements in each row
+def Threshold(A, p):
+    N = A.shape[0]
+    Ap = np.zeros((N, N))
+    for i in range(N):
+        thr = sorted(A[i, :], reverse=True)[p]
+        Ap[i, A[i, :] > thr] = A[i, A[i, :] > thr]
+    return Ap
+
+
+# Computes Laplacian of a matrix
+def Laplacian(A):
+    d = np.sum(A, axis=1) - np.diag(A)
+    D = np.diag(d)
+    return D - A
+
+
+# Calculates eigengaps (differences between adjacent eigenvalues sorted in descending order)
+def Eigengap(S):
+    S = sorted(S)
+    return np.diff(S)
+
+
+# Computes parameters of normalized eigenmaps for automatic thresholding selection
+def ComputeNMEParameters(A, p, max_num_clusters):
+    # p-Neighbour binarization
+    Ap = get_kneighbors_conn(A, p)
+    # Symmetrization
+    Ap = (Ap + np.transpose(Ap)) / 2
+    # Laplacian matrix computation
+    Lp = Laplacian(Ap)
+    # Get max_num_clusters+1 smallest eigenvalues
+    S = scipy.sparse.linalg.eigsh(
+        Lp,
+        k=max_num_clusters + 1,
+        which="SA",
+        tol=1e-6,
+        return_eigenvectors=False,
+        mode="buckling",
+    )
+    # Get largest eigenvalue
+    Smax = scipy.sparse.linalg.eigsh(
+        Lp, k=1, which="LA", tol=1e-6, return_eigenvectors=False, mode="buckling"
+    )
+    # Eigengap computation
+    e = Eigengap(S)
+    g = np.max(e[:max_num_clusters]) / (Smax + 1e-10)
+    r = p / g
+    k = np.argmax(e[:max_num_clusters])
+    return (e, g, k, r)
+
+
+"""
+Performs spectral clustering with Normalized Maximum Eigengap (NME)
+Parameters:
+   A: affinity matrix (matrix of pairwise cosine similarities or PLDA scores between speaker embeddings)
+   num_clusters: number of clusters to generate (if None, determined automatically)
+   max_num_clusters: maximum allowed number of clusters to generate
+   pmax: maximum count for matrix binarization (should be at least 2)
+   pbest: best count for matrix binarization (if 0, determined automatically)
+Returns: cluster assignments for every speaker embedding   
+"""
+
+
+def NME_SpectralClustering(
+    A, num_clusters=None, max_num_clusters=10, pbest=0, pmin=3, pmax=20
+):
+    print(num_clusters,max_num_clusters)
```

Comment:
No print without any string to explain.
This one looks like a debug print. remove it.

Reply:
corrected